client/v3/naming/endpoints/endpoints_impl.go

```diff
@@ -78,73 +77,75 @@ func (m *endpointManager) DeleteEndpoint(ctx context.Context, key string, opts .
 }
 
 func (m *endpointManager) NewWatchChannel(ctx context.Context) (WatchChannel, error) {
-	return nil, fmt.Errorf("Not implemented yet")
-
-	// TODO: Implementation to be inspired by:
-	// Next gets the next set of updates from the etcd resolver.
-	//// Calls to Next should be serialized; concurrent calls are not safe since
-	//// there is no way to reconcile the update ordering.
-	//func (gw *gRPCWatcher) Next() ([]*naming.Update, error) {
-	//	if gw.wch == nil {
-	//		// first Next() returns all addresses
-	//		return gw.firstNext()
-	//	}
-	//	if gw.err != nil {
-	//		return nil, gw.err
-	//	}
-	//
-	//	// process new events on target/*
-	//	wr, ok := <-gw.wch
-	//	if !ok {
-	//		gw.err = status.Error(codes.Unavailable, ErrWatcherClosed.Error())
-	//		return nil, gw.err
-	//	}
-	//	if gw.err = wr.Err(); gw.err != nil {
-	//		return nil, gw.err
-	//	}
-	//
-	//	updates := make([]*naming.Update, 0, len(wr.Events))
-	//	for _, e := range wr.Events {
-	//		var jupdate naming.Update
-	//		var err error
-	//		switch e.Type {
-	//		case etcd.EventTypePut:
-	//			err = json.Unmarshal(e.Kv.Value, &jupdate)
-	//			jupdate.Op = naming.Add
-	//		case etcd.EventTypeDelete:
-	//			err = json.Unmarshal(e.PrevKv.Value, &jupdate)
-	//			jupdate.Op = naming.Delete
-	//		default:
-	//			continue
-	//		}
-	//		if err == nil {
-	//			updates = append(updates, &jupdate)
-	//		}
-	//	}
-	//	return updates, nil
-	//}
-	//
-	//func (gw *gRPCWatcher) firstNext() ([]*naming.Update, error) {
-	//	// Use serialized request so resolution still works if the target etcd
-	//	// server is partitioned away from the quorum.
-	//	resp, err := gw.c.Get(gw.ctx, gw.target, etcd.WithPrefix(), etcd.WithSerializable())
-	//	if gw.err = err; err != nil {
-	//		return nil, err
-	//	}
-	//
-	//	updates := make([]*naming.Update, 0, len(resp.Kvs))
-	//	for _, kv := range resp.Kvs {
-	//		var jupdate naming.Update
-	//		if err := json.Unmarshal(kv.Value, &jupdate); err != nil {
-	//			continue
-	//		}
-	//		updates = append(updates, &jupdate)
-	//	}
-	//
-	//	opts := []etcd.OpOption{etcd.WithRev(resp.Header.Revision + 1), etcd.WithPrefix(), etcd.WithPrevKV()}
-	//	gw.wch = gw.c.Watch(gw.ctx, gw.target, opts...)
-	//	return updates, nil
-	//}
+	resp, err := m.client.Get(ctx, m.target, clientv3.WithPrefix(), clientv3.WithSerializable())
+	if err != nil {
+		return nil, err
+	}
+
+	initUpdates := make([]*Update, 0, len(resp.Kvs))
+	for _, kv := range resp.Kvs {
+		var iup internal.Update
+		if err := json.Unmarshal(kv.Value, &iup); err != nil {
+			continue
```

Comment:
Maybe worth logging the error at least (as WARNING with the 'corrupted' key)

Reply:
Should the logger(`*zap.Logger`) be created internally or passed from outside?

Reply:
Good question. 

1. I assumed clientv3.lg, but this one is private. 
2. logger.go is practically dead  - I will prepare PR to delete this one. 

So there are 2 options: 
  1. Expose the public c.Logger() getter and use it. 
  2. Expand signature of NewManager to take the logger parameter. 
  
@gyuho @jingyih @jpbetz -> do you have opinion here ?

Reply:
OK. Let's go for 1.  It's in scope of the same 'library' and it's hard to imagine use-case when user want's to have separate logger for endpoint resolution. 

Please comment that this method is for internal use of etcd-client library and should not be used as general-purpose logger.

Reply:
OK